### PR TITLE
remove deadlock within the network stack 

### PR DESCRIPTION
**Pull request by stlankes**
To avoid deadlocks, spinlocks and the device checks are revised. In principle, the network thread doesn't block, if another thread is already checking the device.

In addition, the send and receive operations are optimized to stay longer in polling mode and assign the network handler to the core, where the network thread is running. This should reduce the number of wakeup interrupts. This PR depends on hermitcore/libhermit-rs#495

### CODE REVIEW

File: hermit-abi/src/mutex.rs [hunk 1]
```diff
@@ -64,14 +61,31 @@ impl<T> Spinlock<T> {
 		}
 	}
 
+	#[inline]
+	fn obtain_try_lock(&self) -> bool {
```
Comment (mkroening) [inline hermit-abi/src/mutex.rs:65]:
```suggestion
	fn try_obtain_lock(&self) -> bool {
```

File: hermit-sys/src/net/mod.rs [hunk 2]
```diff
@@ -357,9 +347,9 @@ pub(crate) fn network_delay(timestamp: Instant) -> Option<Duration> {
 }
 
 pub(crate) async fn network_run() {
-	future::poll_fn(|cx| match NIC.lock().deref_mut() {
+	future::poll_fn(|_cx| match NIC.lock().deref_mut() {
```
Comment (mkroening) [inline hermit-sys/src/net/mod.rs:350]: Never waking the waker might cause problems. We might have to rethink how this function should work in an async context.

File: hermit-abi/src/mutex.rs (hunk 1 above)
Comment (stlankes) [reply hermit-abi/src/mutex.rs]: I fixed it

File: hermit-sys/src/net/mod.rs (hunk 2 above)
Comment (stlankes) [reply hermit-sys/src/net/mod.rs:361]: Smoltcp is waking the waker.

Comment (mkroening) [reply hermit-sys/src/net/mod.rs:361]: But smoltcp is never getting the waker from `_cx` like this.

Comment (stlankes) [reply hermit-sys/src/net/mod.rs:379]: We register the waker i.e. at https://github.com/hermitcore/rusty-hermit/blob/master/hermit-sys/src/net/mod.rs#L290

**Comment by stlankes** [comment]
I would say that we have to "squash and merge" this PR.

File: hermit-sys/src/net/mod.rs (hunk 2 above)
Comment (mkroening) [reply hermit-sys/src/net/mod.rs:394]: But this `async fn` does not register its waker.

* We could wake the waker immediately, signaling that we want to poll again immediately, but I am not sure if that makes sense for us.
* We could supply the waker to smoltcp, but I am not sure if that is possible.
* We could make this a non-`async` function, as its inner operations are not async as well.

**Comment by mkroening** [comment]
> I would say that we have to "squash and merge" this PR.

Yes. It would be best, if you squashed the commits locally and force pushed. Our bors only supports squash merging as a global configuration.

**Comment by stlankes** [comment]
I will do it

File: hermit-sys/src/net/mod.rs (hunk 2 above)
Comment (stlankes) [reply hermit-sys/src/net/mod.rs:394]: It is a background task. It runs only if blocked_on has time. It is called by run_executor.

Comment (mkroening) [reply hermit-sys/src/net/mod.rs:394]: Ah, I understand. The problem was [`fd23cdd#R31-R42`](https://github.com/hermitcore/rusty-hermit/commit/fd23cddf2d8027ee0fca7396ad5cc03549b283e3#diff-9a7af76af887c12192106d6e158ec152c277af56adcb4a96c6d8bd86755fdb07R31-R42).

That change should be reverted. All futures should take care of waking themselves. I'll create a PR. After that, this function should always wake itself immediately.

Comment (stlankes) [reply hermit-sys/src/net/mod.rs:394]: I reverted already in this PR

Comment (mkroening) [reply hermit-sys/src/net/mod.rs:394]: Where? You can incorporate https://github.com/hermitcore/rusty-hermit/pull/346

Comment (stlankes) [reply hermit-sys/src/net/mod.rs]: Ah, now I understand and revise the PR. Does it fulfill your requirements?

Comment (mkroening) [reply hermit-sys/src/net/mod.rs]: Yes, thank you! :D

**Review by mkroening — APPROVED** [review]
Looks good to me! :)

bors r+

**Pull request merged**